drivers: peripheral UART driver interface overhaul

haukepetersen opened the pull request:
As already merged for GPIO and PR's for SPI (#3216), here is a slight overhaul to the peripheral UART driver. The most notable changes:

- made UART types configurable on a CPU basis
- merged `uart_init_blocking()` and `uart_init()` into one function, dropping the blocking mode
- dropped the `uart_receive_blocking()` function
- changed usage functions to `void xxx(...)`

The main reasons for removing the blocking mode are: (i) the blocking mode is in-efficient and I believe we should not push it's usage by providing this interface and (ii) it was not used anywhere...

For discussion I would like to focus on the changes to the UART interface first. Once we agree on them, I will adjust all the boards/CPUs if neccessary.

gebart commented on boards/f4vi1/include/periph_conf.h:
```diff
@@ -75,24 +82,13 @@ extern "C" {
-#define UART_0_TX_PIN       6
-#define UART_0_RX_PIN       7
-#define UART_0_AF           8
+#define UART_NUMOF          (sizeof(uart_config) / sizeof(uart_conf_t))
```
I suggest `sizeof(uart_config) / sizeof(uart_config[0])` instead

haukepetersen replied:
any particular reasons for this? I don't mind, just try to understand the benefits...

gebart replied:
Not a strong reason, but it doesn't depend on the type of `uart_config` any more.

gebart commented on drivers/include/periph/uart.h:
```diff
@@ -72,95 +87,83 @@ typedef void(*uart_rx_cb_t)(void *arg, char data);
  *
  * @param[in] uart          UART device that will start a transmission
  */
 void uart_tx_begin(uart_t uart);
```
Is this function necessary in the public API?
I think it should be expected that the transmission will begin when we call `uart_write`.

haukepetersen replied:
yes, it is needed. The typical application is

- implement the TX callback so that it takes a byte out of a buffer and calls `uart_write` with that byte
- put some data in a ringbuffer
- call `uart_tx_begin`, which in turn will trigger the TX callback which calls then `uart_write`

gebart commented on drivers/include/periph/uart.h:
```diff
@@ -72,95 +87,83 @@ typedef void(*uart_rx_cb_t)(void *arg, char data);
  *
- * Note: in contrast uart_write, this function will actively wait (block) until the UART
- * device is ready to send a new byte. Consider using the interrupt driven UART mode instead.
+ * In contrast *uart_write*, this function will block (active waiting) until the
```
"in contrast to"

gebart commented:
I like the changes in general. However, I don't understand the detailed usage for TX, how is uart_begin_tx and uart_write supposed to be used together?

Is it possible to change the rx/tx callbacks after initializing the module?
Are there any use cases where it would be desirable to change callbacks without reinitializing?

Will a default TX callback functionality be added to the write syscall handler for stdout?
The stdout update may be related to the file descriptor discussion we had earlier (I forgot the PR/Issue number)

gebart commented:
How should we handle `printf()` from within ISR context?

Newlib may call a flush from inside printf which is expected to be blocking, if we don't block then buffered characters may be lost. 

In particular, it must be possible to print crash information in `core_panic`, even if called from ISR context, before rebooting. (I don't think `core_panic` is actually ever used outside of ISRs)

The printf from inside ISR has caused problems for me in the past.

haukepetersen commented:
TX is indeed the slightly more advanced part of the UART driver. But I see no problem for implementing `printf` or `flush` in a sensible way with the existing interface:

In the syscalls/uart_syscalls implementation, we define a ringbuffer for TX data. When ever `printf` is called, the following happens:

- add data to ringubffer
- call `uart_tx_begin()`

The TX callback should look more or less like this:
```c
int tx_cb(arg *arg)
{
    if (tx_buf.avail > 0) {
        char data = ringbuffer_get_one(&tx_buf);
        uart_write(DEV, data);
        return 1;
    }
    return 0;
```

`flush` could be easily implemented by looping of `while (tx_buf.avail > 0)` with some delay like something ~10ms or so in between.

> Is it possible to change the rx/tx callbacks after initializing the module?
Are there any use cases where it would be desirable to change callbacks without reinitializing?

There is no way at the moment, but neither do I really see a use case for this. But you can of course always call `uart_init` again.

gebart commented:
@haukepetersen What about the case of printf from inside ISR, how do we handle that?

haukepetersen commented:
same, it does not make any difference if you call it from inside an ISR: all printf does is basically filling a ringbuffer and setting the uart TX isr interrupt. All we need to make sure, that we disable interrupts whenever tinkering the the TX buffer, so that there is no concurrent write access to the buffer...

gebart commented:
@haukepetersen What happens if the buffer gets full? Does the printf block? if so, we will get a deadlock if the blocking printf was called from inside an ISR.

haukepetersen commented:
nope, I would not block, Id say we just throw away anything that does not fit into the buffer. Also we could implement both: block, but only if not in interrupt context. This is just a question of how we implement the syscalls, this is completely independent from the UART driver...

gebart commented:
I did a check for PRIMASK in #2605 in blocking mode, maybe some similar solution may be appropriate here?

https://github.com/RIOT-OS/RIOT/pull/2605/files#diff-0f87a847fbd98c6b77134bcd0c723ee5R363

haukepetersen commented:
PRIMASK would not work here, as it is not portable. We already have the `inISR()` function (which calls internally a PRIMASK check on Cortex-Mx platforms), so I would go with that.

gebart commented:
Right, I was confused by the name, since IRQs may be disabled outside of
ISR.
Should the function inISR be renamed "irq_enabled" perhaps? That's a
separate issue though.
On Jun 29, 2015 1:11 PM, "Hauke Petersen" <notifications@github.com> wrote:

> PRIMASK would not work here, as it is not portable. We already have the
> inISR() function (which calls internally a PRIMASK check on Cortex-Mx
> platforms), so I would go with that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/3265#issuecomment-116612284>.
>

haukepetersen commented:
Sure, we could rename it (and do also s/enableIRQ()/irq_enable()/ while we are at it), but that's some other issue...

gebart commented:
PR needs rebase

haukepetersen commented:
rebased.

kaspar030 commented:
> The main reasons for removing the blocking mode are: (i) the blocking mode is in-efficient and I believe we should not push it's usage by providing this interface and (ii) it was not used anywhere...

This (ii) is unfortunately not correct. Our newlib _write() implementation only used blocking mode, and many platforms only implemented that. So essentially, only blocking mode was ever used.

I'm with @gebart here, we need to find a rock-solid concept for ISR and crash cases here, otherwise we'll cause more trouble than the performance gain can justify.

haukepetersen commented:
> This (ii) is unfortunately not correct. Our newlib _write() implementation only used blocking mode, and many platforms only implemented that. So essentially, only blocking mode was ever used.

Almost, the `uart_write_blocking()` is used in the syscalls (and that function is still part of the interface). but the `uart_read_blocking()` was not used, and that one I threw out of the interface...

haukepetersen commented:
> I'm with @gebart here, we need to find a rock-solid concept for ISR and crash cases here, otherwise we'll cause more trouble than the performance gain can justify.

I agree a 100%!

kaspar030 commented:
> but the `uart_read_blocking()` was not used, and that one I threw out of the interface...

Ah ok, perfect!

haukepetersen commented:
rebased.

haukepetersen commented:
addressed typo in doxygen...

haukepetersen commented:
If there is nothing speaking against these changes, I guess it's time to test the effected boards?!

Tested for:
 - [ ] airfy-beacon
 - [ ] arduino-due
 - [ ] arduino-mega2560
 - [ ] cc2538dk
 - [ ] f4vi1
 - [ ] fox
 - [ ] frdm-k64f
 - [ ] iot-lab_M3
 - [ ] mbed_lpc1768
 - [ ] msbiot
 - [ ] mulle
 - [ ] nucleo-f091
 - [ ] nucleo-f303
 - [ ] nucleo-f334
 - [ ] nucleo-l1
 - [ ] openmote
 - [ ] pba-d-01-kw2x
 - [ ] pca10000
 - [ ] pca10005
 - [ ] saml21-xpro
 - [ ] samr21-xpro
 - [ ] spark-core
 - [ ] stm32f0discovery
 - [ ] stm32f3discovery
 - [ ] stm32f4discovery
 - [ ] udoo
 - [ ] yunjia-nrf51822

kaspar030 commented:
If someone decides not to implement non-blocking mode, that would still be possible, right? Or would we have to #ifdef the blocking functions in the interface, or add a define (uart_supports_nonblocking)?

haukepetersen commented:
Partly. Writing to the UART is still possible using a blocking implementation (an so you can skip implementing non-blocking write mode). But receiving is now only possible through an interrupt callback. As this is good style and possible on every platform, I think it is actually a good thing to enforce this.

gebart commented:
I agree with @haukepetersen, it is a reasonable requirement to be able to do reception via ISR.

kaspar030 commented:
Agree too.

daniel-k commented:
It seems like you need to change
```c
#if UART_0_ISR
```
to
```c
#ifdef UART_0_ISR
```
and so on. At least my `gcc version 4.9.3 20150529 (release) [ARM/embedded-4_9-branch revision 224288] (GNU Tools for ARM Embedded Processors)` requires so. Otherwise I get a panic because the dummy handler is called (ISR absent).
The board is the samr21-xpro.

Couldn't test non-blocking yet as the shell captures the first uart, but will do because I really need non-blocking TX now.

haukepetersen commented:
@daniel-k you are right, must have slipped through... Thx for the catch! Will fix it.

gebart commented:
@haukepetersen Needs rebase

haukepetersen commented:
fixed issues pointed out by @daniel-k and rebased.

PeterKietzmann commented:
I can test some boards at the Hack'n'ACK tonight. (Available for me: stm32f4discovery, stm32f0discovery, nucleo-f303, iot-lab_M3, samr21-xpro, arduino-mega2560, arduino-due, yunjia-nrf51822)

A-Paul commented on cpu/sam3/periph/uart.c:
```diff
@@ -48,53 +45,13 @@ static uart_conf_t uart_config[UART_NUMOF];
-    if (res != 0) {
-        return res;
-    }
+    uint16_t clock_divider = F_CPU / (16 * baudrate);
```
Getting an *'unused variable'* warning -treated as error with actual compiler settings- here.

haukepetersen commented:
fixed unused variable for the `sam3`

PeterKietzmann commented:
Does not work on `iot-lab_M3`. Typing "help" in `tests/periph_uart` leads to:
```
2015-07-28 22:10:07,357 - INFO # > *** RIOT kernel panic
2015-07-28 22:10:07,357 - INFO # DUMMY HANDLER
2015-07-28 22:10:07,358 - INFO # *** rebooting...

```

PeterKietzmann commented:
Looks valid on `samr21-xpro`

PeterKietzmann commented:
Does not work on `stm32f0discovery`. Same result as on `iot-lab_M3`

PeterKietzmann commented:
Seems to work on the `nucleo-f303` in general but I can not use the `UART_1`. (`UART_2` works)

PeterKietzmann commented:
STDOUT works on `arduino-mega2560`. Inputs do not work. This is probably related to #3187

haukepetersen commented:
@PeterKietzmann thanks for testing, I will look into it. Seems like this PR has gone bad over time... Will do some more intensive testing (and fixing) this week.

PeterKietzmann commented:
Works on `mbed_lpc1768`

PeterKietzmann commented:
Seems to work on the phyWAVE board

PeterKietzmann commented:
Seems to work on yunjia :-)

PeterKietzmann commented:
@A-Paul did you give it a try on the stm32f4discovery board?

A-Paul commented:
`stm32f4discovery` is doing fine.

haukepetersen commented:
rebased

haukepetersen commented:
f0discovery works fine after rebase.

haukepetersen commented:
fixed the iot-lab, it's now working fine.

haukepetersen commented:
Successfully tested for `airfy-beacon` `f4vi1` `frdm-k64f` `msbiot` `nucleo-f334` `nucleo-l1` `pca10000` and `nrf51dongle`

As these boards were un-touched and other boards with the same CPU were reported to work, I deduct these boards to also be working: `fox` `pca10000` `spark-core` `udoo`

@kaspar030: could you give the `saml21-xpro` a test run?
@thomaseichinger: could try out for the `openmote` (which would also prove the `cc2538dk`)?
@gebart: same for the `mulle`?

This leaves the `nucleo-f091` - which I am not sure where I put it... Will try to find it somewhere around here...

PeterKietzmann commented:
confirmed for iot-lab_M3

haukepetersen commented:
tested and fixed for the `nucleo-f091`, works perfectly now.

gebart commented:
@haukepetersen build fails unrelated to this change on mulle without #3553. After applying the fix this PR works fine.

daniel-k commented:
I tried to port this for the STDIO uart on samr21-xpro, so that the shell an every printf is non-blocking. After some debugging I found the problem that prevented this to work, which rose some questions.

To summarize the concept of TX (or how I understood):
1. write data to ringbuffer
2. manually cause TXC interrupt
3. then in irq_handler write to uart until rungbuffer is empty

But on samd21 in **uart_tx_begin**:
```c
void uart_tx_begin(uart_t dev)
{
    _uart(dev)->INTENSET.reg = SERCOM_USART_INTENSET_TXC;
}
```
you only enable the TXC interrupt. If there hasn't been any writing to `uart->DATA`, the TXC flag is unset and nothing will be transmitted. If TXC was set before everything is fine, the IRQ will be triggered and after the finished transmission you disable the TXC interrupt again.

So how do you ensure that TXC is set in the first place? I didn't find a way to manually trigger the interrupt flag, so writing something at the end of `uart_init` might be a workaround.

```c
    /* ... */
    NVIC_EnableIRQ(SERCOM0_IRQn + _sercom_id(_uart(dev)));
    uart->INTENSET.reg = SERCOM_USART_INTENSET_RXC;
    uart->DATA.reg = (uint8_t)0;
    return 0;
}
```

But maybe I missed something? I didn't test an additional uart on samr21-xpro yet.

gebart commented:
needs rebase

haukepetersen commented:
Rebased and introduced one more optimization: I changed the transmit sequence slightly: When sending the non-blocking way, the byte to send is not written to the device using the `uart_write()` method anymore, but it is returned by the TX callback. This makes the TX mode even more easy to use and understand...

haukepetersen commented:
@daniel-k: looking into the samd now...

haukepetersen commented:
@daniel-k: Just tried you example, with these two changes:
```c
static uint16_t uart_stdio_tx_cb(void *arg)
{
    int c = ringbuffer_get_one(&_tx_buf);
    if (c == -1) {
        return 0;
    }
    return (uint16_t)((c & 0xff) | UART_TX_MORE_DATA);
}
```
and 
```c
int uart_stdio_write(const char* buffer, int len)
{
    ringbuffer_add(&_tx_buf, buffer, (unsigned)len);
    uart_tx(STDIO);
    return len;
}
```
As far as I can tell the UART driver works just fine... Can you re-check with the newest changes in this PR?

gebart commented:
I'm seeing some strange corruption on mulle during boot with this branch.
It does not seem to appear after boot however.

```
2015-08-24 13:26:55,889 - INFO # > kernel_init(): This is RIOT! (Version: 2014.12-2915-g0464-haukepetersen-opt_samd21_uart)
���rrRtU�k֫��3:26:57,536 - INFO # kernel_init(): jumping into first �
            �U��U�!�kernel_init(): This is RIOT! (Version: 2014.12-2915-g0464-haukepetersen-opt_samd21_uart)
2015-08-24 13:26:57,539 - INFO # kernel_init(): jumping into first �ask...
2015-08-24 13:26:57,541 - INFO # Welcome to RIOT!
2015-08-24 13:26:59,550 - INFO # > kernel_init(): This is RIOT! (Version: 2014.12-2915-g0464-haukepetersen-opt_samd21_uart)
2015-08-24 13:26:59,554 - INFO # kernel_init(): jumping into first �ask...
2015-08-24 13:26:59,555 - INFO # Welcome to RIOT!
2015-08-24 13:27:01,546 - INFO # > kernel_init(): This is RIOT! (Version: 2014.12-2915-g0464-haukepetersen-opt_samd21_uart)
2015-08-24 13:27:02,673 - INFO # kernel_init(): jumping into first ���rrRtU�k֫��
                                                                               �U��I��kernel_init(): This is RIOT! (Version: 2014.12-2915-g0464-haukepetersen-opt_samd21_uart)
2015-08-24 13:27:02,676 - INFO # kernel_init(): jumping into first ��sk...
2015-08-24 13:27:02,678 - INFO # Welcome to RIOT!
2015-08-24 13:27:03,971 - INFO # > kernel_init(): This is RIOT! (Version: 2014.12-2915-g0464-haukepetersen-opt_samd21_uart)
2015-08-24 13:27:04,868 - INFO # kernel_init(): jumping into first �    ���rrRtU�k֫��
                                                                                    �U��U���kernel_init(): This is RIOT! (Version: 2014.12-2915-g0464-haukepetersen-opt_samd21_uart)
2015-08-24 13:27:04,872 - INFO # kernel_init(): jumping into first �ask...
2015-08-24 13:27:04,874 - INFO # Welcome to RIOT!
2015-08-24 13:27:05,901 - INFO # > kernel_init(): This is RIOT! (Version: 2014.12-2915-g0464-haukepetersen-opt_samd21_uart)
2015-08-24 13:27:05,905 - INFO # kernel_init(): jumping into first �ask...
2015-08-24 13:27:05,906 - INFO # Welcome to RIOT!
2015-08-24 13:27:06,866 - INFO # > kernel_init(): This is RIOT! (Version: 2014.12-2915-g0464-haukepetersen-opt_samd21_uart)
2015-08-24 13:27:07,820 - INFO # kernel_init(): jumping into first !    ���rrRtU�k֫��
                                                                                    �U�U�!�kernel_init(): This is RIOT! (Version: 2014.12-2915-g0464-haukepetersen-opt_samd21_uart)
2015-08-24 13:27:09,738 - INFO # kernel_init(): jumping into first !
                                                                    �m�rrRtU�k֫��
                                                                                �U��I��kernel_init(): This is RIOT! (Version: 2014.12-2915-g0464-haukepetersen-opt_samd21_uart)
2015-08-24 13:27:10,629 - INFO # kernel_init(): jumping into first �    �-�rrRtU�k֫��
                                                                                    �U�U�!�kernel_init(): This is RIOT! (Version: 2014.12-2915-g0464-haukepetersen-opt_samd21_uart)
2015-08-24 13:27:11,576 - INFO # kernel_init(): jumping into first �    ���rrRtU�k֫��
                                                                                    �U��U�!�kernel_init(): This is RIOT! (Version: 2014.12-2915-g0464-haukepetersen-opt_samd21_uart)
2015-08-24 13:27:11,579 - INFO # kernel_init(): jumping into first �ask...
2015-08-24 13:27:11,581 - INFO # Welcome to RIOT!
2015-08-24 13:27:12,886 - INFO # > kernel_init(): This is RIOT! (Version: 2014.12-2915-g0464-haukepetersen-opt_samd21_uart)
2015-08-24 13:27:13,809 - INFO # kernel_init(): jumping into first �    �m�rrRtU�k֫��
                                                                                    �U��U�!�kernel_init(): This is RIOT! (Version: 2014.12-2915-g0464-haukepetersen-opt_samd21_uart)
2015-08-24 13:27:13,812 - INFO # kernel_init(): jumping into first task...
2015-08-24 13:27:13,814 - INFO # Welcome to RIOT!
```

haukepetersen commented:
Look strange indeed. I try if I can reproduce this behavior on some other board...

haukepetersen commented:
I just tried with the Phytec `pba-d-01-kw2x` board, and everything runs fine there. So I am wondering what could cause this... The UART on `mulle` is running with a baudrate of 115200 and you are using pyterm to interact with the board, right?

haukepetersen commented:
oh, and rebased.

daniel-k commented:
@haukepetersen 
I rechecked with the newest changes and your proposed adaptions and it seems to work fine, but still only with adding `uart->DATA.reg = (uint8_t)0;` add the end of `uart_init(...)`. 

Now that you changed the way `uart_tx_begin(...)` works, my (maybe better) [workaround](https://github.com/daniel-k/RIOT/commit/a5be02018ea3d99ef674fceaa2732d260a397110) won't work anymore.

haukepetersen commented:
Sorry for the late reply, busy times...

I cant really reproduce that behavior here, I will try to look into it once more. Would be nice to get this PR in at some time :-)

haukepetersen commented:
After some more test and discussions offline, I updated the UART interface once more. I decided for a different concept when sending data. The main reasons for this are:
- sending data using DMA was not possible, as the interface forced the use of interrupt driven sending or active waiting
- the old concept was hard to understand by people who just wanted to use the interface
- I run into trouble on many platforms due to erratas on the handling of the TX interrupt bit

Now there is only once send function `uart_write`. This function get's a buffer and the length as arguments and only returns, once the buffer is send. It is then up to the low-level driver implementation, how this sending is handled. This may be with active waiting, interrupt driven or using DMA - all depends how much effort is spend on the driver...

A test implementation is done for the `stm32f4` -> `stm32f4discovery`

@daniel-k, @PeterKietzmann, @gebart, @kaspar030: What to you think about this change? If you approve, I will finally go through all the CPUs again and adapt them.

daniel-k commented:
> Now there is only once send function uart_write. This function get's a buffer and the length as arguments and only returns, once the buffer is send.

I haven't looked at any implementation yet, just at `periph/uart.h` that basically has the same wording: `once the buffer is send` If take this literally it would mean, that any `uart_write`, no matter which mechanism (DMA, IRQ, blocking) is used, would block the calling thread for long times, just imagine low baud rates. Did you maybe mean to say `once the buffer is consumed`, i.e. the buffer is copied somewhere, e.g. DMA takes care of sending, and control is transfered back to the caller? This would make much more sense to me.

haukepetersen commented:
No, I actually menat 'once the buffer is send'. The idea is, that the function really does block for the time it takes to send the data and only returns after the bytes have been send out. This does however not implay, that the thread is actively blocking, as it is doing when using the current `uart_send_blocking`. When switching to interrupt driven or better DMA, it would mean that the function waits on a mutex or similar, allowing the CPU to go to sleep, until the data is send. 

Making the send function synchronous as this reduces the implementation complexity by some factors and removes on many occasions the need for double buffering, hence making the use of UART more memory efficient.

daniel-k commented:
I see your point. The use case I had recently was that I had **a lot** debug output in some relatively timing sensitive code that could only be debugged reasonably with non-blocking UART. I'm not sure if there are others.
Of course you can implement a debugging thread then that buffers the output, but this increases complexity and overhead by more than you save here I guess.

kaspar030 commented:
> What to you think about this change?

I like it a lot. The interface is probably as simple as possible now.

OlegHahm commented:
Also I like the proposal in general, I think @daniel-k has a very valid point here.

(And to be a smartass, the documentation should be `once the buffer is sent` (or `has been sent`)).

kaspar030 commented:
> I think @daniel-k has a very valid point here.

We should introduce caching, but not at this level.

haukepetersen commented:
@daniel-k I see your use-case. But in general, I don't see `printf` as the vehicle of choice when it comes to debugging time-sensitive code! There are ways far more suited for debugging this kind of things, e.g. Logic analyzers + gpio pins or similar...

> We should introduce caching, but not at this level.

Yes! This was always the idea with this interface.

daniel-k commented:
@haukepetersen 
Sure debugging with GPIOs is nice for getting the timing right, but not to get internal states of your code. There are sophisticated debug tools for ARM, but unfortunately there's no free toolchain and cheap debugger that supports them. 

In my case the code breaks when the debugging introduces too big delays, but I can't just debug with 2 GPIOs. A trace tool would be the right choice.

OlegHahm commented:
And not everyone has everywhere access to logic analyzers. Think for, e.g. testbeds.

kaspar030 commented:
The issue is not whether we want fast uart output, but if it should become asynchronous at this level (within the uart interface).

OlegHahm commented:
Yes, I agree with this.

OlegHahm commented:
So, basically we need an accompanying PR that introduces this caching somewhere above, right?

kaspar030 commented:
@OlegHahm Wasn't all uart output synchronous before?

OlegHahm commented:
Was it? I dunno. If this is true, yes, we can work on this later on.

daniel-k commented:
I think @kaspar030 is right. But @haukepetersen was introducing interrupt-driven UART with this PR that got me excited and now you take that toy away again :-D

haukepetersen commented:
> Was it? 

YES.

> But @haukepetersen was introducing interrupt-driven UART with this PR

Not quite. Interrupt driven UART did exist before... And to clarify - I am not taking it away again, the decision of how to use the UART is simply moved from the user of the interface into the implementation of the interface.

sgso commented:
> This function is blocking, as it will only return after @p len bytes from the given buffer have been sen**t**

For UARTs with integrated FIFO buffers, I assume it's ok to just pass the data to it and return? (and check on the next call, if the FIFO buffer can be written again?)

haukepetersen commented:
> For UARTs with integrated FIFO buffers, I assume it's ok to just pass the data to it and return? (and check on the next call, if the FIFO buffer can be written again?)

Yes.

haukepetersen commented:
closed in favor of #4114
